src/Optical/Emitters/Directions.jl

```diff
@@ -0,0 +1,206 @@
+module Directions
+
+using ....OpticSim, ...Geometry
+using ...Emitters
+using LinearAlgebra
+
+abstract type AbstractDirectionDistribution{T<:Real} end
+
+#---------------------------------------
+# Direction Distrubution Common Utilities
+#---------------------------------------
+
+export generate
+
+Base.iterate(a::AbstractDirectionDistribution, state = 1) = state > length(a) ? nothing : (generate(a, state - 1), state + 1)
+Base.getindex(a::AbstractDirectionDistribution, index) = generate(a, index)
+Base.firstindex(a::AbstractDirectionDistribution) = 0
+Base.lastindex(a::AbstractDirectionDistribution) = length(a) - 1
+Base.copy(a::AbstractDirectionDistribution) = a # most don't have any heap allocated stuff so don't really need copying
+
+
+#---------------------------------------
+# Constant Ray Direction
+#---------------------------------------
+"""
+Encapsulates a single ray direction, where the default direction is unitZ3 [0, 0, 1].
+
+```julia
+Constant(direction::Vec3{T}) where {T<:Real}
+Constant(::Type{T} = Float64) where {T<:Real}
+```
+"""
+struct Constant{T} <: AbstractDirectionDistribution{T}
+    direction::Vec3{T}
+
+    function Constant(direction::Vec3{T}) where {T<:Real}
+        return new{T}(direction)
+    end
+    
+    function Constant(::Type{T} = Float64) where {T<:Real}
+        direction = unitZ3(T)
+        return new{T}(direction)
+    end
+end
+export Constant
+
+Base.length(d::Constant{T}) where {T} = 1
+function Emitters.generate(d::Constant{T}, n::Int) where {T<:Real}
```

Comment:
It seems strange to define an Emitters function inside the Directions module. This is going to make it hard for users to find all relevant function definitions, since they will be scattered in unlikely places. 

Why is this necessary?

Reply:
From my understanding, this is a good solution if we are going with the submodule approach.

Essentially, Emitters has "generate()" in its interface, and this can be called within various contexts, such as for a hexapolar origin or composite source.

Reply:
Sorry, misread the code. This seems okay given that we want modules nested so deeply.